staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -385,6 +385,9 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 			runtime.HandleError(fmt.Errorf("error copying from remote stream to local connection: %v", err))
 		}
 
+		// drain the remote datastream to make sure we don't leave it blocked
+		_, _ = io.Copy(io.Discard, dataStream)
```

Comment:
should we only do this if the io.Copy returned an error above?

Reply:
I was surprised the spdy library didn't drain the streams by itself and forced the users to drain manually, that is why I commented in https://github.com/kubernetes/kubernetes/pull/126718#issuecomment-2296801100, at least in http2 reset frames have higher priority and are processed before data frame, so you are able to tear down the connection despite the buffer is full.


The problem seems to be we are using `defer dataStream.Close()`  instead of `defer dataStream.Reset()` to tear down the stream, 

https://github.com/kubernetes/kubernetes/blob/22a30e7cbbee31b51da2f5e88833efbfad2c3f6b/vendor/github.com/moby/spdystream/stream.go#L169-L186

Close() leave the stream in a half close state, and I think we want to completely close it, actually we are using Reset() for closing the connection

https://github.com/kubernetes/kubernetes/blob/22a30e7cbbee31b51da2f5e88833efbfad2c3f6b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection.go#L107-L112

If you apply the following patch, replacing Close() by Reset() the test passes too and I think is more correct

```diff
diff --git a/staging/src/k8s.io/client-go/tools/portforward/portforward.go b/staging/src/k8s.io/client-go/tools/portforward/portforward.go
index 3b9dc26c623..762880f6fa8 100644
--- a/staging/src/k8s.io/client-go/tools/portforward/portforward.go
+++ b/staging/src/k8s.io/client-go/tools/portforward/portforward.go
@@ -385,16 +385,13 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
                        runtime.HandleError(fmt.Errorf("error copying from remote stream to local connection: %v", err))
                }

-               // drain the remote datastream to make sure we don't leave it blocked
-               _, _ = io.Copy(io.Discard, dataStream)
-
                // inform the select below that the remote copy is done
                close(remoteDone)
        }()

        go func() {
                // inform server we're not sending any more data after copy unblocks
-               defer dataStream.Close()
+               defer dataStream.Reset()

                // Copy from the local port to the remote side.
                if _, err := io.Copy(dataStream, conn); err != nil && !strings.Contains(err.Error(), "use of closed network connection") {

```

Reply:
Your explanation is excellent. After reading the code snippet you posted, I understand that the `Reset` function is similar to the RST flag in the TCP protocol and can forcibly terminate a spdy stream. Although spdy streams are full-duplex, it doesn't make sense to keep a one-way spdy stream when the local connection of kubectl port-forward has already been disconnected because there is no client to process the received data (it can only be discarded by kubectl). Therefore, I think replacing `Close` with `Reset` is indeed reasonable.